compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -174,8 +174,8 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             .param_env
             .caller_bounds()
             .iter()
-            .filter_map(|p| p.to_opt_poly_trait_pred())
-            .filter(|p| !p.references_error());
+            .filter(|p| !p.references_error())
+            .filter_map(|p| p.to_opt_poly_trait_pred());
```

Comment:
The reason this is faster is because `Predicate`s cache their type flags, whereas `PolyTraitPredicate`s do not. So gathering the type flags for a poly trait predicate is `O(# of substs)` and gathering the type flags for a predicate is `O(1)`.